### ISSUE TITLE
Literal class property downcasted to a base type on usage

### ISSUE THREAD

**Issue by outring** [microsoft/pyright#11189]
**Describe the bug**

A `Literal` property of a class received via `__init__` loses its literal type info and appears to be a plain `str` when used in class methods

**Code or Screenshots**

```python
from typing import Literal

TextMatching = Literal["equals", "contains", "starts", "ends"]

def match_text(actual: str, expected: str, matching: TextMatching) -> bool:
    return True

class TextConstraint:
    def __init__(self, expected: str, matching: TextMatching):
        self._expected = expected
        self._matching = matching

    def validate(self, actual: str) -> bool:
        return match_text(actual, self._expected, self._matching)
    #                                             ^ Argument of type "str" cannot be assigned to parameter "matching" of type "TextMatching" in function "match_text"
```

Notes:
* It can be worked around by adding explicit type to an assignment `self._matching: TextMatching = matching`
* It preserves `Literal` type in the `__init__`, i.e. `match_text(actual, self._expected, self._matching)` won't produce an error in the constructor

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug.

You haven't provided a type declaration for the instance variable `_matching`, so its type is inferred based on all values assigned to it within the class. In your code sample, there is only one such assignment where it is being assigned a value whose type is `Literal["equals", "contains", "starts", "ends"]`. Literal types are widened to their non-literal types during inference. For example, if you added the statement `self._foo = 1`, the type of `_foo` would be inferred as `int`, not as `Literal[1]`. 

If your intent is for `_matching` to be constrained to a literal type, you should declare it explicitly as such.

```python
        self._matching: TextMatching = matching
```
